omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,899 @@
+
+                // Sort neighbours.
+                const auto size = valid_neighbours.size();
+                sort_by_degree(&valid_neighbours[0], size, degrees);
+
+                // Write the processed neighbours.
+                const auto base_write_offset =
+                    base_offset + next_level_idx + write_offset;
+
+                // Memcpy would be nice, but we need element-wise atomicity.
+                for (auto i = 0; i < size; ++i) {
+#pragma omp atomic write
+                    perm[base_write_offset + i] = valid_neighbours[i];
+                }
+
+                write_offset += size;
+                valid_neighbours.clear();
+            }
+        }
+    }
+}
+
+
+/**
+ * Processes all isolated nodes, returning their count.
+ */
+template <typename IndexType>
+IndexType handle_isolated_nodes(std::shared_ptr<const OmpExecutor> exec,
+                                const IndexType *const row_ptrs,
+                                const IndexType *const col_idxs,
+                                const IndexType *const degrees,
+                                IndexType *const perm, size_type num_vertices,
+                                gko::vector<uint8> &previous_component)
+{
+    struct IsolatedNodes {
+        // Using a gko::vector here makes clang and gcc segfault,
+        // if we then use 'omp_priv = omp_orig'.
+        std::vector<IndexType> nodes;
+        IsolatedNodes &operator+=(const IsolatedNodes &rhs)
+        {
+            nodes.reserve(nodes.size() + rhs.nodes.size());
+            nodes.insert(nodes.end(), rhs.nodes.begin(), rhs.nodes.end());
+            return *this;
+        }
+    };
+
+#pragma omp declare reduction(FindIsolated : IsolatedNodes : omp_out += omp_in)
+    auto isolated = IsolatedNodes();
+#pragma omp parallel for reduction(FindIsolated : isolated)
+    for (auto i = 0; i < num_vertices; ++i) {
+        // No need to check for diagonal elements (only self-neighbouring) here,
+        // those are already removed from the matrix.
+        if (degrees[i] == 0) {
+            isolated.nodes.push_back(i);
+            previous_component[i] = true;
+        }
+    }
+
+    std::memcpy(perm, &isolated.nodes[0],
+                isolated.nodes.size() * sizeof(IndexType));
+    return isolated.nodes.size();
+}
+
+
+/**
+ * Computes a rcm permutation, employing the parallel unordered rcm algorithm.
+ */
+template <typename IndexType>
+void get_permutation(std::shared_ptr<const OmpExecutor> exec,
+                     const size_type num_vertices,
+                     const IndexType *const row_ptrs,
+                     const IndexType *const col_idxs,
+                     const IndexType *const degrees, IndexType *const perm,
+                     IndexType *const inv_perm,
+                     const gko::reorder::starting_strategy strategy)
+{
+    // Initialize the perm to all "signal value".
+    std::fill(perm, perm + num_vertices, std::numeric_limits<IndexType>::max());
```

Comment:
I can see your concern with all these special values for internal signaling. I will write a short paragraph at the start of the function about the meaning of signal values in different containers thorough the algorithm. 
In this specific case, the signal values are rather critical for performance. As the permutation is written concurrently, it is a nice advantage to only need to read one value atomically to know 1. If that value has been written yet, and 2. if yes, what the written value is. It comes at the expense of one unit in the value space, that is true. That should maybe be documented.

Reply:
I'm not at all against sentinel values, I would just recommend using a value that can't possibly occur normally in the permutation (like -1, since IndexType is always signed) instead of a value that could occur for huge matrices.

Reply:
Guess I was still too caught up in my implementation here, which uses unsigned values wherever possible, also for the permutation. Yes, that should be a -1 then.